scipy/spatial/spherical_voronoi.py

```diff
@@ -0,0 +1,416 @@
+    #seems to be relatively rare in my own work, but sufficiently common to cause crashes when iterating over large amounts of messy data
+    if scipy.spatial.distance.pdist(array_ordered_Voronoi_polygon_vertices).min() < (10 ** -7):
+        return 10 ** -8
+    else:
+        n = array_ordered_Voronoi_polygon_vertices.shape[0]
+        #point we start from
+        root_point = array_ordered_Voronoi_polygon_vertices[0]
+        totalexcess = 0
+        #loop from 1 to n-2, with point 2 to n-1 as other vertex of triangle
+        # this could definitely be written more nicely
+        b_point = array_ordered_Voronoi_polygon_vertices[1]
+        root_b_dist = calculate_haversine_distance_between_spherical_points(root_point, b_point, 1.0)
+        for i in 1 + np.arange(n - 2):
+            a_point = b_point
+            b_point = array_ordered_Voronoi_polygon_vertices[i+1]
+            root_a_dist = root_b_dist
+            root_b_dist = calculate_haversine_distance_between_spherical_points(root_point, b_point, 1.0)
+            a_b_dist = calculate_haversine_distance_between_spherical_points(a_point, b_point, 1.0)
+            s = (root_a_dist + root_b_dist + a_b_dist) / 2
+            totalexcess += 4 * math.atan(math.sqrt( math.tan(0.5 * s) * math.tan(0.5 * (s-root_a_dist)) * math.tan(0.5 * (s-root_b_dist)) * math.tan(0.5 * (s-a_b_dist))))
+        return totalexcess * (sphere_radius ** 2)
+
+def calculate_surface_area_of_planar_polygon_in_3D_space(array_ordered_Voronoi_polygon_vertices):
+    '''Based largely on: http://stackoverflow.com/a/12653810
+    Use this function when spherical polygon surface area calculation fails (i.e., lots of nearly-coplanar vertices and negative surface area).'''
+    #unit normal vector of plane defined by points a, b, and c
+    def unit_normal(a, b, c):
+        x = np.linalg.det([[1,a[1],a[2]],
+             [1,b[1],b[2]],
+             [1,c[1],c[2]]])
+        y = np.linalg.det([[a[0],1,a[2]],
+             [b[0],1,b[2]],
+             [c[0],1,c[2]]])
+        z = np.linalg.det([[a[0],a[1],1],
+             [b[0],b[1],1],
+             [c[0],c[1],1]])
+        magnitude = (x**2 + y**2 + z**2)**.5
+        return (x/magnitude, y/magnitude, z/magnitude)
+
+    #area of polygon poly
+    def poly_area(poly):
+        '''Accepts a list of xyz tuples.'''
+        assert len(poly) >= 3, "Not a polygon (< 3 vertices)."
+        total = [0, 0, 0]
+        N = len(poly)
+        for i in range(N):
+            vi1 = poly[i]
+            vi2 = poly[(i+1) % N]
+            prod = np.cross(vi1, vi2)
+            total[0] += prod[0]
+            total[1] += prod[1]
+            total[2] += prod[2]
+        result = np.dot(total, unit_normal(poly[0], poly[1], poly[2]))
+        return abs(result/2)
+
+    list_vertices = [] #need a list of tuples for above function
+    for coord in array_ordered_Voronoi_polygon_vertices:
+        list_vertices.append(tuple(coord))
```

Comment:
Could use `list(map(tuple, array_ordered_Voronoi_polygon_vertices))` here to avoid the loop.